### ISSUE TITLE
Rule S5944: VB.NET functions should use the return statement

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2159]
### Description
In Visual Basic, it allowed by to assign the return value to a variable (that does not have to be declared) with the same name of the function, instead of using a return statement. That is a bad practice that should be discouraged.

```vb
Public Class UseReturnStatement

	Public Function FunctionName() As Integer
		FunctionName = 42 ' Not compliant
	End Function

	Public Function ReturnStatement() As Integer
		Return 42 ' Compliant
	End Function

	Public Function ReturnWithVariable() As Integer
		Dim value As Integer = 42
		Return value  ' compliant
	End Function


	Public Function ReturnWithFunctionName() As Integer
		ReturnWithFunctionName= 42
		Return ReturnWithFunctionName  ' Not compliant
	End Function

End Class
```

**Comment by Corniel** [comment]
As I can not create a pull request:

This piece of code should do the trick. I'm not sure if it is faster to start with an identifier and that go way back, or start with an function block.

```CSharp
using System;
using System.Collections.Immutable;
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.VisualBasic;
using Microsoft.CodeAnalysis.VisualBasic.Syntax;
using SonarAnalyzer.Common;
using SonarAnalyzer.Helpers;

namespace SonarAnalyzer.Rules.VisualBasic
{
    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
    [Rule(DiagnosticId)]
    public sealed class FunctionShouldReturnExplictly : SonarDiagnosticAnalyzer
    {
        internal const string DiagnosticId = "S???";
        private const string MessageFormat = "Do not use the VB construct of assigning a return statement via the function name variable, but use an explicit return statement.";
        private static readonly DiagnosticDescriptor rule =
            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);

        protected override void Initialize(SonarAnalysisContext context)
        {
            context.RegisterSyntaxNodeActionInNonGenerated(c =>
            {
                var name = (IdentifierNameSyntax)c.Node;

                if (name.Ancestors()?.FirstOrDefault(IsFunctionBlockSyntax) is MethodBlockSyntax methodBlock)
                {
                    var statement = (MethodStatementSyntax)methodBlock.BlockStatement;

                    if (name.Identifier.ValueText.Equals(statement?.Identifier.ValueText, StringComparison.InvariantCultureIgnoreCase))
                    {
                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, name.GetLocation()));
                    }
                }
            },
            SyntaxKind.IdentifierName);
        }

        private bool IsFunctionBlockSyntax(SyntaxNode node)
        {
            return node is MethodBlockSyntax methodBlock &&
                methodBlock.BlockStatement.DeclarationKeyword.IsKind(SyntaxKind.FunctionKeyword);
        }
    }
}
```

**Comment by Corniel** [comment]
Any udate on this?

**Comment by pavel-mikula-sonarsource** [comment]
The rule is now specified as [RSPEC-5944](https://jira.sonarsource.com/browse/RSPEC-5944). @nicolas-harraudeau-sonarsource  Can you please take a look?

**Comment by nicolas-harraudeau-sonarsource** [comment]
@pavel-mikula-sonarsource All good. Just did a few minor changes, including a time to fix of 5min and adding tags "confusing" and "bad-practice". Feel free to change them.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel,

I've created scaffolding for this rule in `Pavel/Corniel-S5944-Return` branch.

Since we'd like to have the issue raised only once in the function and add all other cases as secondary locations, it would be good to start with the function declaration and search the descendant nodes for assignment.

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource I'll try to look into it tomorrow!

**Comment by Corniel** [comment]
I'll look into this tomorrow.
